server/src/main/java/org/elasticsearch/cluster/coordination/Publication.java

```diff
@@ -0,0 +1,376 @@
+    }
+
+    private void onPossibleCompletion() {
+        if (isCompleted) {
+            return;
+        }
+
+        if (timedOut == false) {
+            for (final PublicationTarget target : publicationTargets) {
+                if (target.publicationTargetStateMachine.isActive()) {
+                    return;
+                }
+            }
+        }
+
+        if (applyCommitRequest.isPresent() == false) {
+            logger.debug("onPossibleCompletion: [{}] commit failed", this);
+            assert isCompleted == false;
+            isCompleted = true;
+            onCompletion(false);
+            return;
+        }
+
+        assert isCompleted == false;
+        isCompleted = true;
+        onCompletion(true);
+        assert applyCommitRequest.isPresent();
+        logger.trace("onPossibleCompletion: [{}] was successful, applying new state locally", this);
+    }
+
+    // For assertions only: verify that this invariant holds
+    private boolean publicationCompletedIffAllTargetsInactiveOrTimedOut() {
+        if (timedOut == false) {
+            for (final PublicationTarget target : publicationTargets) {
+                if (target.publicationTargetStateMachine.isActive()) {
+                    return isCompleted == false;
+                }
+            }
+        }
+        return isCompleted;
+    }
+
+    private void onPossibleCommitFailure() {
+        if (applyCommitRequest.isPresent()) {
+            onPossibleCompletion();
+            return;
+        }
+
+        final CoordinationState.VoteCollection possiblySuccessfulNodes = new CoordinationState.VoteCollection();
+        for (PublicationTarget publicationTarget : publicationTargets) {
+            if (publicationTarget.publicationTargetStateMachine.mayCommitInFuture()) {
+                possiblySuccessfulNodes.addVote(publicationTarget.discoveryNode);
+            } else {
+                assert publicationTarget.publicationTargetStateMachine.isFailed() : publicationTarget.publicationTargetStateMachine;
+            }
+        }
+
+        if (isPublishQuorum(possiblySuccessfulNodes) == false) {
+            logger.debug("onPossibleCommitFailure: non-failed nodes {} do not form a quorum, so {} cannot succeed",
+                possiblySuccessfulNodes, this);
+            Exception e = new Discovery.FailedToCommitClusterStateException("non-failed nodes do not form a quorum");
+            publicationTargets.stream().filter(PublicationTarget::isActive).forEach(pt -> pt.setFailed(e));
+            onPossibleCompletion();
+        }
+    }
+
+    protected abstract void onCompletion(boolean committed);
+
+    protected abstract boolean isPublishQuorum(CoordinationState.VoteCollection votes);
+
+    protected abstract Optional<ApplyCommitRequest> handlePublishResponse(DiscoveryNode sourceNode, PublishResponse publishResponse);
+
+    protected abstract void onPossibleJoin(DiscoveryNode sourceNode, PublishWithJoinResponse response);
```

Comment:
Does this need the whole `PublishWithJoinResponse` or could it just take the `Optional<Join>`? Or maybe even the `Join` only if present?

Reply:
I don't know yet. I've added a TODO in 43e3399f164